lib/java/com/google/android/material/tabs/TabLayout.java

```diff
@@ -2644,7 +2669,9 @@ public boolean performClick() {
         if (!handled) {
           playSoundEffect(SoundEffectConstants.CLICK);
         }
-        tab.select();
+        if (tab.onTabClickListener != null) {
+          tab.onTabClickListener.onClick(tab);
+        }
```

Comment:
Let's try to avoid overriding the default behavior unless there's a valid use case to do that.

Reply:
@drchen Hey, thanks for the comment.

The main idea is to have an option to override a default behavior.
A usecase is described in the https://github.com/material-components/material-components-android/issues/4105 issue.

tl.dr.
We need to override a default behaviour to add a validation logic when user clicks on the tab before selecting it.
With the existing implementation, I tried selecting the previous tab in case the data is not valid but it doesn't look good because the tab is selected with the animation.

Reply:
Ok cool. So I think there are several improvements I'll suggest here:

1. Let's make `onTabClickListener`  be non-null.
2. Let `OnTabClickListener .onClick()` return a boolean to indicate if the click event is handled, so we can simplify the logic like:
```
final boolean handled = super.performClick() || tab.onTabClickListener.onClick(tab);
... rest the same
```
3. Refine the javadoc to make it more clear that how to override the default behavior.